econml/policy/forest/_forest.py

```diff
@@ -0,0 +1,326 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+#
+# This code contains snippets of code from
+# https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/ensemble/_forest.py
+# published under the following license and copyright:
+# BSD 3-Clause License
+#
+# Copyright (c) 2007-2020 The scikit-learn developers.
+# All rights reserved.
+
+import numbers
+from warnings import catch_warnings, simplefilter, warn
+from abc import ABCMeta, abstractmethod
+import numpy as np
+import threading
+from ..._ensemble import (BaseEnsemble, _partition_estimators, _get_n_samples_subsample, _accumulate_prediction)
+from ...utilities import check_inputs, cross_product
+from ...tree._tree import DTYPE, DOUBLE
+from ._tree import PolicyTree
+from joblib import Parallel, delayed
+from scipy.sparse import hstack as sparse_hstack
+from sklearn.utils import check_random_state, compute_sample_weight
+from sklearn.utils.validation import _check_sample_weight, check_is_fitted
+from sklearn.utils import check_X_y
+
+__all__ = ["PolicyForest"]
+
+MAX_INT = np.iinfo(np.int32).max
+
+# =============================================================================
+# Policy Forest
+# =============================================================================
+
+
+class PolicyForest(BaseEnsemble, metaclass=ABCMeta):

```

Comment:
Same here, a lot of similarities with the GRF forest, except that the GRF has a lot more auxiliary functions (about 2/3 of the code). Most of the components here appear in the GRF forest as well.

Reply:
I tried for this one, but there was so much different small details that I thought it was ok for the forest.

I merged them in the tree, but left the forests separate.